File: sdks/go/leaf/registry.go
// Package leaf provides leaf node registration and auto-adaptation.
package leaf

import (
	stdctx "context"
	"encoding/json"
	"fmt"
	"reflect"
	"runtime"
	"strings"
	"sync"

	icecontext "github.com/zjn-zjn/ice/sdks/go/context"
	"github.com/zjn-zjn/ice/sdks/go/dto"
	"github.com/zjn-zjn/ice/sdks/go/enum"
	"github.com/zjn-zjn/ice/sdks/go/node"
	"github.com/zjn-zjn/ice/sdks/go/scan"
)

// LeafMeta contains metadata about a leaf node.
type LeafMeta struct {
	Name  string
	Desc  string
	Order int
	// Alias provides alternative names for this leaf class.
	// Useful for multi-language compatibility (e.g., Java/Go/Python using different class names).
	Alias []string
}

// AfterPropertiesSet is an optional interface that leaf nodes can implement
// to perform initialization after all properties have been set.
// This is called after JSON configuration is applied to the node.
type AfterPropertiesSet interface {
	AfterPropertiesSet()
}

// LeafErrorHandler is an optional interface that leaf nodes can implement
// to provide custom error handling logic when an error occurs during execution.
// The returned RunState determines the behavior:
//   - SHUT_DOWN: re-panic the error and stop execution
//   - TRUE/FALSE/NONE: use this state and continue execution
type LeafErrorHandler interface {
	ErrorHandle(ctx stdctx.Context, roam *icecontext.Roam, err error) enum.RunState
}

// LeafFlow is implemented by flow-type leaf nodes.
type LeafFlow interface {
	DoFlow(ctx stdctx.Context, roam *icecontext.Roam) bool
}

// LeafResult is implemented by result-type leaf nodes.
type LeafResult interface {
	DoResult(ctx stdctx.Context, roam *icecontext.Roam) bool
}

// LeafNone is implemented by none-type leaf nodes.
type LeafNone interface {
	DoNone(ctx stdctx.Context, roam *icecontext.Roam)
}

// leafEntry holds registration info for a leaf type.
type leafEntry struct {
	meta       *LeafMeta
	factory    func() any
	nodeType   enum.NodeType
	iceFields  []dto.IceFieldInfo
	hideFields []dto.IceFieldInfo
	roamKeys   []dto.RoamKeyMeta
}

// LeafNode is a node that can have its base properties set.
type LeafNode interface {
	node.Node
	node.BaseAccessor
	SetBase(base *node.Base)
}

var (
	registry      = make(map[string]*leafEntry)
	aliasRegistry = make(map[string]string) // alias -> className
	roamKeysStore = make(map[string][]dto.RoamKeyMeta) // className -> roamKeys (populated by generated code)
	mu            sync.RWMutex
)

// SetRoamKeys sets the roam key metadata for a given class name.
// This is typically called from generated code (roam_keys_gen.go) during init().
func SetRoamKeys(className string, keys []dto.RoamKeyMeta) {
	mu.Lock()
	defer mu.Unlock()
	roamKeysStore[className] = keys
	// Also update existing entry if already registered
	if entry, ok := registry[className]; ok {
		entry.roamKeys = keys
	}
}

// ApplyRoamKeysMap applies a map of short struct names to roam key metadata.
// It matches short names (e.g. "ScoreFlow") to registered full class names
// (e.g. "com.ice.test.flow.ScoreFlow") by suffix.
func ApplyRoamKeysMap(m map[string][]dto.RoamKeyMeta) {
	mu.Lock()
	defer mu.Unlock()
	for shortName, keys := range m {
		suffix := "." + shortName
		for fullName, entry := range registry {
			if strings.HasSuffix(fullName, suffix) || fullName == shortName {
				roamKeysStore[fullName] = keys
				entry.roamKeys = keys
				break
			}
		}
	}
}

// Register registers a leaf node factory with the given class name.
// The factory should return a pointer to a struct that implements one of:
// - LeafFlow (for flow type)
// - LeafResult (for result type)
// - LeafNone (for none type)
//
// Fields are automatically extracted from the struct using reflection:
// - Fields with `ice:"name:xxx,desc:xxx"` tag become iceFields (visible in UI)
// - Other fields with `json` tag become hideFields
//
// If meta.Alias is provided, those names will also be registered as aliases
// pointing to this class name.
func Register(className string, meta *LeafMeta, factory func() any) {
	if className == "" || factory == nil {
		return
	}

	// Create a sample to detect the type and extract fields
	sample := factory()
	nodeType := detectNodeType(sample)
	iceFields, hideFields := extractFields(sample)

	// Use pre-registered roam keys (from generated code), or auto-scan from source
	roamKeys := roamKeysStore[className]
	if len(roamKeys) == 0 {
		roamKeys = autoScanRoamKeys(sample)
	}

	mu.Lock()
	defer mu.Unlock()
	registry[className] = &leafEntry{
		meta:       meta,
		factory:    factory,
		nodeType:   nodeType,
		iceFields:  iceFields,
		hideFields: hideFields,
		roamKeys:   roamKeys,
	}

	// Register aliases
	if meta != nil {
		for _, alias := range meta.Alias {
			if alias != "" && alias != className {
				aliasRegistry[alias] = className
			}
		}
	}
}

// autoScanRoamKeys locates the source file where the leaf's business method
// (DoFlow/DoResult/DoNone) is defined via runtime reflection, then AST-scans it.
// Returns nil silently if source is unavailable (e.g., production deployment).
func autoScanRoamKeys(sample any) []dto.RoamKeyMeta {
	t := reflect.TypeOf(sample)
	file := findMethodSourceFile(t, "DoFlow", "DoResult", "DoNone")
	if file == "" {
		return nil
	}

	results, err := scan.ScanFile(file)
	if err != nil {
		return nil
	}

	if t.Kind() == reflect.Ptr {
		t = t.Elem()
	}
	structName := t.Name()

	for _, r := range results {
		if r.ClassName == structName {
			return r.RoamKeys
		}
	}
	return nil
}

// findMethodSourceFile returns the source file path where one of the named methods is defined.
func findMethodSourceFile(t reflect.Type, methodNames ...string) string {
	for _, name := range methodNames {
		m, ok := t.MethodByName(name)
		if !ok {
			continue
		}
		pc := m.Func.Pointer()
		fn := runtime.FuncForPC(pc)
		if fn == nil {
			continue
		}
		file, _ := fn.FileLine(pc)
		return file
	}
	return ""
}

// ResolveClassName resolves a config name to the actual class name.
// If confName is an alias, returns the real class name.
// Otherwise returns confName unchanged.
func ResolveClassName(confName string) string {
	mu.RLock()
	defer mu.RUnlock()
	if realName, ok := aliasRegistry[confName]; ok {
		return realName
	}
	return confName
}

// extractFields extracts field information from a struct using reflection.
// Fields with `ice` tag become iceFields, others become hideFields.
func extractFields(sample any) (iceFields, hideFields []dto.IceFieldInfo) {
	v := reflect.ValueOf(sample)
	if v.Kind() == reflect.Ptr {
		v = v.Elem()
	}
	if v.Kind() != reflect.Struct {
		return nil, nil
	}

	t := v.Type()
	for i := 0; i < t.NumField(); i++ {
		field := t.Field(i)

		// Skip unexported fields
		if !field.IsExported() {
			continue
		}

		// Get json field name
		jsonTag := field.Tag.Get("json")
		if jsonTag == "" || jsonTag == "-" {
			continue
		}
		jsonName := strings.Split(jsonTag, ",")[0]

		// Get ice tag for iceFields
		iceTag := field.Tag.Get("ice")
		if iceTag == "-" {
			continue // ice:"-" means completely ignore this field
		}
		if iceTag != "" {
			// Parse ice tag: ice:"name:xxx,desc:xxx"
			fieldInfo := dto.IceFieldInfo{
				Field: jsonName,
				Type:  getTypeName(field.Type),
			}
			for _, part := range strings.Split(iceTag, ",") {
				kv := strings.SplitN(part, ":", 2)
				if len(kv) == 2 {
					switch kv[0] {
					case "name":
						fieldInfo.Name = kv[1]
					case "desc":
						fieldInfo.Desc = kv[1]
					}
				}
			}
			iceFields = append(iceFields, fieldInfo)
		} else {
			// hideField
			hideFields = append(hideFields, dto.IceFieldInfo{
				Field: jsonName,
				Type:  getTypeName(field.Type),
			})
		}
	}
	return
}

// getTypeName returns a language-agnostic type name.
func getTypeName(t reflect.Type) string {
	switch t.Kind() {
	case reflect.String:
		return "string"
	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32:
		return "int"
	case reflect.Int64:
		return "long"
	case reflect.Float32:
		return "float"
	case reflect.Float64:
		return "double"
	case reflect.Bool:
		return "boolean"
	case reflect.Slice, reflect.Array:
		return "list"
	case reflect.Map:
		return "map"
	default:
		return "object"
	}
}

// detectNodeType detects the node type based on the interfaces implemented.
func detectNodeType(leaf any) enum.NodeType {
	if _, ok := leaf.(LeafFlow); ok {
		return enum.TypeLeafFlow
	}
	if _, ok := leaf.(LeafResult); ok {
		return enum.TypeLeafResult
	}
	if _, ok := leaf.(LeafNone); ok {
		return enum.TypeLeafNone
	}
	return enum.TypeLeafNone
}

// CreateNode creates a leaf node from configuration.
// confName can be a class name or an alias.
func CreateNode(confName, confField string) (node.Node, error) {
	mu.RLock()
	// First, try to resolve alias
	className := confName
	if realName, ok := aliasRegistry[confName]; ok {
		className = realName
	}
	entry, ok := registry[className]
	mu.RUnlock()

	if !ok {
		return nil, fmt.Errorf("leaf class not found: %s (resolved: %s)", confName, className)
	}

	// Create instance
	instance := entry.factory()

	// Apply configuration
	if confField != "" && confField != "{}" {
		if err := json.Unmarshal([]byte(confField), instance); err != nil {
			return nil, fmt.Errorf("failed to unmarshal leaf config: %w", err)
		}
	}

	// Call AfterPropertiesSet if implemented
	if aps, ok := instance.(AfterPropertiesSet); ok {
		aps.AfterPropertiesSet()
	}

	// Wrap with appropriate adapter
	return wrapLeaf(instance, entry.nodeType), nil
}

// wrapLeaf wraps a leaf instance with the appropriate node wrapper.
func wrapLeaf(instance any, nodeType enum.NodeType) node.Node {
	switch nodeType {
	case enum.TypeLeafFlow:
		return &flowLeafNode{instance: instance}
	case enum.TypeLeafResult:
		return &resultLeafNode{instance: instance}
	case enum.TypeLeafNone:
		return &noneLeafNode{instance: instance}
	default:
		return &noneLeafNode{instance: instance}
	}
}

// flowLeafNode wraps a flow-type leaf.
type flowLeafNode struct {
	node.Leaf
	instance any
}

// SetBase sets the base properties for this leaf node.
func (f *flowLeafNode) SetBase(base *node.Base) {
	f.Base = *base
}

func (f *flowLeafNode) Process(ctx stdctx.Context, roam *icecontext.Roam) enum.RunState {
	return node.ProcessWithBase(ctx, &f.Base, roam, f.doLeaf, f.getErrorHandler())
}

// getErrorHandler returns the error handler if the instance implements LeafErrorHandler.
func (f *flowLeafNode) getErrorHandler() node.ErrorHandler {
	if handler, ok := f.instance.(LeafErrorHandler); ok {
		return &leafErrorHandlerAdapter{handler: handler}
	}
	return nil
}

func (f *flowLeafNode) doLeaf(ctx stdctx.Context, roam *icecontext.Roam) enum.RunState {
	if leaf, ok := f.instance.(LeafFlow); ok {
		if leaf.DoFlow(ctx, roam) {
			return enum.TRUE
		}
		return enum.FALSE
	}
	return enum.FALSE
}

// leafErrorHandlerAdapter adapts LeafErrorHandler to node.ErrorHandler.
type leafErrorHandlerAdapter struct {
	handler LeafErrorHandler
}

func (a *leafErrorHandlerAdapter) ErrorHandle(ctx stdctx.Context, roam *icecontext.Roam, err error) enum.RunState {
	return a.handler.ErrorHandle(ctx, roam, err)
}

// resultLeafNode wraps a result-type leaf.
type resultLeafNode struct {
	node.Leaf
	instance any
}

// SetBase sets the base properties for this leaf node.
func (r *resultLeafNode) SetBase(base *node.Base) {
	r.Base = *base
}

func (r *resultLeafNode) Process(ctx stdctx.Context, roam *icecontext.Roam) enum.RunState {
	return node.ProcessWithBase(ctx, &r.Base, roam, r.doLeaf, r.getErrorHandler())
}

// getErrorHandler returns the error handler if the instance implements LeafErrorHandler.
func (r *resultLeafNode) getErrorHandler() node.ErrorHandler {
	if handler, ok := r.instance.(LeafErrorHandler); ok {
		return &leafErrorHandlerAdapter{handler: handler}
	}
	return nil
}

func (r *resultLeafNode) doLeaf(ctx stdctx.Context, roam *icecontext.Roam) enum.RunState {
	if leaf, ok := r.instance.(LeafResult); ok {
		if leaf.DoResult(ctx, roam) {
			return enum.TRUE
		}
		return enum.FALSE
	}
	return enum.FALSE
}

// noneLeafNode wraps a none-type leaf.
type noneLeafNode struct {
	node.Leaf
	instance any
}

// SetBase sets the base properties for this leaf node.
func (n *noneLeafNode) SetBase(base *node.Base) {
	n.Base = *base
}

func (n *noneLeafNode) Process(ctx stdctx.Context, roam *icecontext.Roam) enum.RunState {
	return node.ProcessWithBase(ctx, &n.Base, roam, n.doLeaf, n.getErrorHandler())
}

// getErrorHandler returns the error handler if the instance implements LeafErrorHandler.
func (n *noneLeafNode) getErrorHandler() node.ErrorHandler {
	if handler, ok := n.instance.(LeafErrorHandler); ok {
		return &leafErrorHandlerAdapter{handler: handler}
	}
	return nil
}

func (n *noneLeafNode) doLeaf(ctx stdctx.Context, roam *icecontext.Roam) enum.RunState {
	if leaf, ok := n.instance.(LeafNone); ok {
		leaf.DoNone(ctx, roam)
	}
	return enum.NONE
}

// GetLeafNodes returns information about all registered leaf nodes.
func GetLeafNodes() []dto.LeafNodeInfo {
	mu.RLock()
	defer mu.RUnlock()

	result := make([]dto.LeafNodeInfo, 0, len(registry))
	for className, entry := range registry {
		info := dto.LeafNodeInfo{
			Type:       byte(entry.nodeType),
			Clazz:      className,
			Order:      100,
			IceFields:  entry.iceFields,
			HideFields: entry.hideFields,
			RoamKeys:   entry.roamKeys,
		}
		if entry.meta != nil {
			info.Name = entry.meta.Name
			info.Desc = entry.meta.Desc
			if entry.meta.Order > 0 {
				info.Order = entry.meta.Order
			}
		}
		result = append(result, info)
	}
	return result
}

// IsRegistered checks if a leaf class is registered (including aliases).
func IsRegistered(className string) bool {
	mu.RLock()
	defer mu.RUnlock()
	// Check if it's directly registered
	if _, ok := registry[className]; ok {
		return true
	}
	// Check if it's an alias
	if realName, ok := aliasRegistry[className]; ok {
		_, exists := registry[realName]
		return exists
	}
	return false
}

// GetAliasRegistry returns a copy of the alias registry (for debugging).
func GetAliasRegistry() map[string]string {
	mu.RLock()
	defer mu.RUnlock()
	result := make(map[string]string, len(aliasRegistry))
	for k, v := range aliasRegistry {
		result[k] = v
	}
	return result
}

File: sdks/python/src/ice/leaf/registry.py
"""Leaf node registration and auto-adaptation for Ice SDK."""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Any, Callable, Type, get_type_hints, get_origin, get_args

from ice.dto import LeafNodeInfo, IceFieldInfo
from ice.enums import NodeType, RunState
from ice.node.base import Base, Node, process_with_base
from ice.node.leaf import Leaf

# For Annotated support
try:
    from typing import Annotated, get_origin, get_args
except ImportError:
    from typing_extensions import Annotated, get_origin, get_args

if TYPE_CHECKING:
    from ice.context.context import Context
    from ice.context.pack import Pack
    from ice.context.roam import Roam


@dataclass
class IceField:
    """
    Field metadata for ice fields.
    
    Use with typing.Annotated to mark fields as ice fields (visible in UI):
    
        class ScoreFlow:
            score: Annotated[int, IceField(name="分数阈值", desc="判断分数")] = 0
            key: Annotated[str, IceField(name="键名")] = "score"
    """
    name: str = ""
    desc: str = ""


@dataclass
class IceIgnore:
    """
    Marker to ignore a field completely (not in iceFields or hideFields).
    
    Use with typing.Annotated:
    
        class MyNode:
            _service: SomeService  # Already ignored by _ prefix
            cache: Annotated[dict, IceIgnore()] = None  # Explicitly ignored
    """
    pass


# Alias for backward compatibility
FieldMeta = IceField


@dataclass
class LeafMeta:
    """Metadata for a leaf node."""
    name: str = ""
    desc: str = ""
    order: int = 100
    # Aliases for multi-language compatibility
    # When config's confName matches an alias, it maps to this class
    alias: list[str] = field(default_factory=list)
    # Optional: explicitly define iceFields with metadata (legacy way)
    ice_fields: dict[str, IceField] | None = None


@dataclass
class LeafEntry:
    """Registration entry for a leaf type."""
    cls: Type
    meta: LeafMeta | None
    node_type: NodeType
    ice_fields: list[IceFieldInfo]
    hide_fields: list[IceFieldInfo]


# Global registry: class_name -> LeafEntry
_registry: dict[str, LeafEntry] = {}
# Alias registry: alias -> class_name
_alias_registry: dict[str, str] = {}


def _get_type_name(py_type: Any) -> str:
    """Convert Python type to language-agnostic type name."""
    # Handle Annotated types - extract the actual type
    origin = get_origin(py_type)
    if origin is Annotated:
        args = get_args(py_type)
        if args:
            py_type = args[0]
            origin = get_origin(py_type)
    
    if py_type is str:
        return "string"
    elif py_type is int:
        return "long"
    elif py_type is float:
        return "double"
    elif py_type is bool:
        return "boolean"
    elif origin is list:
        return "list"
    elif origin is dict:
        return "map"
    return "object"


def _extract_fields(cls: Type, meta: LeafMeta | None) -> tuple[list[IceFieldInfo], list[IceFieldInfo]]:
    """
    Extract field information from a class using annotations.
    
    Supports two ways to define ice fields:
    1. Using Annotated: `score: Annotated[int, IceField(name="xxx")]`
    2. Using meta.ice_fields dict (legacy)
    """
    ice_fields: list[IceFieldInfo] = []
    hide_fields: list[IceFieldInfo] = []
    
    # Get type hints including Annotated metadata
    try:
        hints = get_type_hints(cls, include_extras=True)
    except Exception:
        hints = getattr(cls, "__annotations__", {})
    
    # Get legacy iceFields metadata from LeafMeta if provided
    legacy_ice_fields = meta.ice_fields if meta and meta.ice_fields else {}
    
    for field_name, field_type in hints.items():
        # Skip private fields
        if field_name.startswith("_"):
            continue
        
        # Check if this is an Annotated type with IceField or IceIgnore
        ice_field_meta: IceField | None = None
        is_ignored = False
        actual_type = field_type
        
        origin = get_origin(field_type)
        if origin is Annotated:
            args = get_args(field_type)
            if args:
                actual_type = args[0]
                # Look for IceField or IceIgnore in the annotations
                for arg in args[1:]:
                    if isinstance(arg, IceIgnore):
                        is_ignored = True
                        break
                    if isinstance(arg, IceField):
                        ice_field_meta = arg
        
        # Skip ignored fields
        if is_ignored:
            continue
        
        # Fall back to legacy meta.ice_fields
        if ice_field_meta is None and field_name in legacy_ice_fields:
            ice_field_meta = legacy_ice_fields[field_name]
        
        type_name = _get_type_name(actual_type)
        
        if ice_field_meta is not None:
            # This is an iceField
            ice_fields.append(IceFieldInfo(
                field=field_name,
                name=ice_field_meta.name,
                desc=ice_field_meta.desc,
                type=type_name,
            ))
        else:
            # This is a hideField
            hide_fields.append(IceFieldInfo(
                field=field_name,
                type=type_name,
            ))
    
    return ice_fields, hide_fields


def _detect_node_type(cls: Type) -> NodeType:
    """Detect the node type based on which methods the class implements."""
    # Check Flow types
    if hasattr(cls, "do_roam_flow") or hasattr(cls, "do_pack_flow") or hasattr(cls, "do_flow"):
        return NodeType.LEAF_FLOW
    # Check Result types
    if hasattr(cls, "do_roam_result") or hasattr(cls, "do_pack_result") or hasattr(cls, "do_result"):
        return NodeType.LEAF_RESULT
    # Check None types
    if hasattr(cls, "do_roam_none") or hasattr(cls, "do_pack_none") or hasattr(cls, "do_none"):
        return NodeType.LEAF_NONE
    # Default to None type
    return NodeType.LEAF_NONE


def leaf(
    class_name: str,
    *,
    name: str = "",
    desc: str = "",
    order: int = 100,
    alias: list[str] | str | None = None,
    ice_fields: dict[str, IceField] | None = None,
) -> Callable[[Type], Type]:
    """
    Decorator to register a leaf node class.
    
    Usage with Annotated (recommended):
        @ice.leaf("com.example.ScoreFlow", name="分数判断", alias=["score_flow"])
        class ScoreFlow:
            score: Annotated[int, IceField(name="分数阈值", desc="判断分数")] = 0
            key: Annotated[str, IceField(name="键名")] = "score"
            
            def do_roam_flow(self, roam: Roam) -> bool:
                return roam.get_int(self.key, 0) >= self.score
    
    Args:
        class_name: The fully qualified class name (used for matching with config)
        name: Display name for the node
        desc: Description of the node
        order: Display order in UI
        alias: Alias names for multi-language compatibility (can be string or list)
        ice_fields: Optional dict mapping field names to IceField (legacy way)
    
    Returns:
        The decorated class (unchanged)
    """
    # Normalize alias to list
    alias_list: list[str] = []
    if alias:
        if isinstance(alias, str):
            alias_list = [alias]
        else:
            alias_list = list(alias)
    
    def decorator(cls: Type) -> Type:
        meta = LeafMeta(name=name, desc=desc, order=order, alias=alias_list, ice_fields=ice_fields)
        node_type = _detect_node_type(cls)
        extracted_ice_fields, hide_fields = _extract_fields(cls, meta)
        
        entry = LeafEntry(
            cls=cls,
            meta=meta,
            node_type=node_type,
            ice_fields=extracted_ice_fields,
            hide_fields=hide_fields,
        )
        
        # Register main class name
        _registry[class_name] = entry
        
        # Register aliases
        for a in alias_list:
            if a and a != class_name:
                _alias_registry[a] = class_name
        
        return cls
    return decorator


def register_leaf(
    class_name: str,
    cls: Type,
    meta: LeafMeta | None = None,
) -> None:
    """
    Manually register a leaf node class.
    
    Args:
        class_name: The fully qualified class name
        cls: The class to register
        meta: Optional metadata (including alias)
    """
    node_type = _detect_node_type(cls)
    ice_fields, hide_fields = _extract_fields(cls, meta)
    
    entry = LeafEntry(
        cls=cls,
        meta=meta,
        node_type=node_type,
        ice_fields=ice_fields,
        hide_fields=hide_fields,
    )
    
    _registry[class_name] = entry
    
    # Register aliases
    if meta and meta.alias:
        for a in meta.alias:
            if a and a != class_name:
                _alias_registry[a] = class_name


def resolve_class_name(conf_name: str) -> str:
    """
    Resolve a config name to the actual class name.
    If conf_name is an alias, returns the real class name.
    Otherwise returns conf_name unchanged.
    """
    return _alias_registry.get(conf_name, conf_name)


def is_registered(class_name: str) -> bool:
    """Check if a leaf class is registered (including aliases)."""
    resolved = resolve_class_name(class_name)
    return resolved in _registry


def get_leaf_nodes() -> list[LeafNodeInfo]:
    """Get information about all registered leaf nodes."""
    result = []
    for class_name, entry in _registry.items():
        info = LeafNodeInfo(
            type=entry.node_type.value,
            clazz=class_name,
            order=100,
            iceFields=entry.ice_fields,
            hideFields=entry.hide_fields,
        )
        if entry.meta:
            info.name = entry.meta.name
            info.desc = entry.meta.desc
            if entry.meta.order > 0:
                info.order = entry.meta.order
        result.append(info)
    return result


def get_alias_registry() -> dict[str, str]:
    """Get the alias registry (for debugging)."""
    return _alias_registry.copy()


def create_leaf_node(conf_name: str, conf_field: str) -> Node:
    """
    Create a leaf node from configuration.
    
    Args:
        conf_name: The class name to create (can be alias)
        conf_field: JSON configuration for the node
    
    Returns:
        A Node instance wrapping the leaf
    
    Raises:
        ValueError: If the class is not registered
    """
    # Resolve alias to real class name
    class_name = resolve_class_name(conf_name)
    
    if class_name not in _registry:
        raise ValueError(f"Leaf class not found: {conf_name} (resolved: {class_name})")
    
    entry = _registry[class_name]
    
    # Create instance
    instance = entry.cls()
    
    # Apply configuration
    if conf_field and conf_field != "{}":
        try:
            config = json.loads(conf_field)
            for key, value in config.items():
                if hasattr(instance, key):
                    setattr(instance, key, value)
        except json.JSONDecodeError:
            pass
    
    # Wrap with appropriate adapter
    return _wrap_leaf(instance, entry.node_type)


def _wrap_leaf(instance: Any, node_type: NodeType) -> Node:
    """Wrap a leaf instance with the appropriate node wrapper."""
    if node_type == NodeType.LEAF_FLOW:
        return FlowLeafNode(instance)
    elif node_type == NodeType.LEAF_RESULT:
        return ResultLeafNode(instance)
    else:
        return NoneLeafNode(instance)


class FlowLeafNode(Leaf, Node):
    """Wrapper for flow-type leaf nodes."""
    
    def __init__(self, instance: Any) -> None:
        super().__init__()
        self._instance = instance
    
    def process(self, ctx: Context) -> RunState:
        return process_with_base(self, ctx, self._do_leaf)
    
    def _do_leaf(self, ctx: Context) -> RunState:
        result = False
        
        # Try most specific first (Roam), then Pack, then Context
        if hasattr(self._instance, "do_roam_flow"):
            result = self._instance.do_roam_flow(ctx.pack.roam)
        elif hasattr(self._instance, "do_pack_flow"):
            result = self._instance.do_pack_flow(ctx.pack)
        elif hasattr(self._instance, "do_flow"):
            result = self._instance.do_flow(ctx)
        
        return RunState.TRUE if result else RunState.FALSE


class ResultLeafNode(Leaf, Node):
    """Wrapper for result-type leaf nodes."""
    
    def __init__(self, instance: Any) -> None:
        super().__init__()
        self._instance = instance
    
    def process(self, ctx: Context) -> RunState:
        return process_with_base(self, ctx, self._do_leaf)
    
    def _do_leaf(self, ctx: Context) -> RunState:
        result = False
        
        if hasattr(self._instance, "do_roam_result"):
            result = self._instance.do_roam_result(ctx.pack.roam)
        elif hasattr(self._instance, "do_pack_result"):
            result = self._instance.do_pack_result(ctx.pack)
        elif hasattr(self._instance, "do_result"):
            result = self._instance.do_result(ctx)
        
        return RunState.TRUE if result else RunState.FALSE


class NoneLeafNode(Leaf, Node):
    """Wrapper for none-type leaf nodes."""
    
    def __init__(self, instance: Any) -> None:
        super().__init__()
        self._instance = instance
    
    def process(self, ctx: Context) -> RunState:
        return process_with_base(self, ctx, self._do_leaf)
    
    def _do_leaf(self, ctx: Context) -> RunState:
        if hasattr(self._instance, "do_roam_none"):
            self._instance.do_roam_none(ctx.pack.roam)
        elif hasattr(self._instance, "do_pack_none"):
            self._instance.do_pack_none(ctx.pack)
        elif hasattr(self._instance, "do_none"):
            self._instance.do_none(ctx)
        
        return RunState.NONE


def set_leaf_base(node: Node, base: Base) -> None:
    """Set the base properties on a leaf node wrapper."""
    if isinstance(node, (FlowLeafNode, ResultLeafNode, NoneLeafNode)):
        # Copy base properties
        node.ice_node_id = base.ice_node_id
        node.ice_time_type = base.ice_time_type
        node.ice_start = base.ice_start
        node.ice_end = base.ice_end
        node.ice_node_debug = base.ice_node_debug
        node.ice_inverse = base.ice_inverse
        node.ice_forward = base.ice_forward
        node.ice_error_state = base.ice_error_state
        node.ice_log_name = base.ice_log_name
        node.ice_type = base.ice_type
